validation/elliptic_solvers/doubly_bounded_poisson.jl

```diff
@@ -0,0 +1,197 @@
+    event = launch!(child_arch, grid, :xyz, ∇²!, ∇²φ, grid, φ, dependencies=Event(device(child_arch)))
+    wait(device(child_arch), event)
+    fill_halo_regions!(∇²φ)
+
+    return nothing
+end
+
+# Solve ∇²φ = r with `FFTBasedPoissonSolver`
+φ_fft = CenterField(grid)
+fft_solver = FFTBasedPoissonSolver(grid)
+fft_solver.storage .= interior(r)
+
+@info "Solving the Poisson equation with an FFT-based solver..."
+@time solve!(φ_fft, fft_solver, fft_solver.storage)
+
+fill_halo_regions!(φ_fft)
+
+
+# Solve ∇²φ = r with `PreconditionedConjugateGradientSolver`
+φ_cg = CenterField(grid)
+cg_solver = PreconditionedConjugateGradientSolver(compute_∇²!, template_field=r, reltol=eps(eltype(grid)))
+
+@info "Solving the Poisson equation with a conjugate gradient iterative solver..."
+@time solve!(φ_cg, cg_solver, r, arch, grid)
+
+fill_halo_regions!(φ_cg)
+
+
+# Solve ∇²φ = r with `AlgebraicMultigrid` solver
+function create_matrix(grid, linear_operator!, args...)
+    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
+    # Currently assuming Nz = 1
+    A = spzeros(eltype(grid), Nx*Ny, Nx*Ny)
+
+    make_column(f) = reshape(interior(f), (Nx*Ny*Nz, 1))
+
+    eᵢⱼₖ = CenterField(grid)
+    ∇²eᵢⱼₖ = CenterField(grid)
+    for k in 1:Nz, j in 1:Ny, i in 1:Nx
+        eᵢⱼₖ .= 0
+        ∇²eᵢⱼₖ .= 0
+        eᵢⱼₖ[i, j, k] = 1
+        fill_halo_regions!(eᵢⱼₖ)
+
+        linear_operator!(∇²eᵢⱼₖ, eᵢⱼₖ, args...)
+
+        A[:, Nx*(j-1)+i] = make_column(∇²eᵢⱼₖ)
+    end
+    return A
+end
+
+
+@info "Solving the Poisson equation with the Algebraic Multigrid solver..."
+A = create_matrix(grid, compute_∇²!, arch, grid)
+
+r_array = array_type(arch)(reshape(interior(r), Nx * Ny * Nz))
+
+ml = ruge_stuben(A, maxiter=100)
+n = length(ml) == 1 ? size(ml.final_A, 1) : size(ml.levels[1].A, 1)
+V = promote_type(eltype(ml.workspace), eltype(r_array))
+φ_mg_array = zeros(V, size(r_array))
+@show @allocated _solve!(φ_mg_array, ml, r_array)
+# @info "Solving the Poisson equation with the Algebraic Multigrid solver (not in-place)..."
+# @time φ_mg_array = solve(A, r_array, RugeStubenAMG(), maxiter=100, verbose=true)
+
+φ_mg = CenterField(grid)
+interior(φ_mg) .= reshape(φ_mg_array, Nx, Ny, Nz)
+fill_halo_regions!(φ_mg)
+
+
+# Solve ∇²φ = r with `PreconditionedConjugateGradientSolver` solver using the AlgebraicMultigrid as preconditioner
+
+struct MultigridPreconditioner{M, A}
+    matrix_operator :: M
+    maxiter :: Int
+    amg_algorithm :: A
+end
+
+mgp = MultigridPreconditioner(A, 5, RugeStubenAMG())
+
+"""
+    precondition!(z, mgp::MultigridPreconditioner, r, args...)
+
+Return `z` (Field)
+"""
+function precondition!(z, mgp::MultigridPreconditioner, r, args...)
+    Nx, Ny, Nz = r.grid.Nx, r.grid.Ny, r.grid.Nz
+    
+    r_array = array_type(arch)(reshape(interior(r), Nx * Ny * Nz))
+
+    z_array = solve(mgp.matrix_operator, r_array, mgp.amg_algorithm, maxiter=mgp.maxiter)
+
+    interior(z) .= reshape(z_array, Nx, Ny, Nz)
+    fill_halo_regions!(z)
+
+    return z
+end
```

Comment:
@elise-palethorpe, we should convert this to a non-allocating one :)

Reply:
would something like this

```julia
_solve!(z_array, mgp.amg_algorithm, r_array, maxiter=mgp.maxiter)
```

do the job?

perhaps we need to use `z` somehow so that we don't allocate `z_array`?